### PR TITLE
Vni manager

### PR DESCRIPTION
**Pull request by glenna-m**
Please review vni-manager.js & its associated test file vni-manager-mocha.js

### CODE REVIEW

File: src/input-states.js [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+ * @param state a State object containing both a data payload and an lm.
+ *
+ * @return getter returns the current input port state state.
+ *         setter returns the current context
+ *
+ * Usage to retrieve state:
+ *  inputState(node, vnid) : {portName: currentState, addressablePortName: [currentState]}
+ *  inputState(node, vnid, portName, socketId) : currentState
+ * Usage to store state:
+ *  inputState(node, vnid, {portName: currentState, addressablePortName: [currentState]}) : this
```
Comment (dbooth-boston) [inline src/input-states.js]: Please add a JSdoc comment about @this: will it be the node or the inputStates facade?

File: src/input-states.js [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the input port whose state is being recorded
+ * @param socketId index of the socket or undefined
+ * @param state a State object containing both a data payload and an lm.
+ *
+ * @return getter returns the current input port state state.
+ *         setter returns the current context
+ *
+ * Usage to retrieve state:
+ *  inputState(node, vnid) : {portName: currentState, addressablePortName: [currentState]}
```
Comment (dbooth-boston) [inline src/input-states.js:18]: Let's use the term 'multi' instead of 'addressable' in RDF Pipeline code (though NoFlo uses the term 'addressable').

Why is this only for a multi port?

File: src/input-states.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+/**
+ * Returns a hash of port names to port states or array of states if the port is
+ * addressable/multi. This provides the caller with all portNames, distingishes between
+ * addressable/multi and non-addressable/multi ports, and provides the caller with all their
+ * states that will be useful when calling an updater.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ */
+function getAllPortStates(node, vnid) {
```
Comment (dbooth-boston) [inline src/input-states.js:46]: good

File: src/input-states.js [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+/**
+ * Returns the state of the port or an array of states, if the port is
+ * addressable/multi. When all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the InPort name on this node 
+ */
+function getPortState(node, vnid, portName) {
```
Comment (dbooth-boston) [inline src/input-states.js:61]: good

File: src/input-states.js [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+/**
+ * Returns an array of the port states, one item for each attached socket. When
+ * all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the InPort name on this node 
+ */
+function getPortStateArray(node, vnid, portName) {
```
Comment (dbooth-boston) [inline src/input-states.js:79]: good

File: src/input-states.js [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+ * look up the last state when checking to see if the state has changed.
+ * Note that the socketId is not neccessarily the same as the array index
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the name of the addressable/multi port for this node
+ * @param socketId the socketId provided to a port data event
+ */
+function getPortStateBySocketId(node, vnid, portName, socketId) {
+    if (socketId == null)
+        return getPortState(node, vnid, portName);
```
Comment (dbooth-boston) [inline src/input-states.js:98]: I think it would be safer to throw an error in this case.  The caller should have called getPortState if there is no socketId.

File: src/input-states.js [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+/**
+ * Changes the state of each given port. If a port name is missing it is unmodified.
+ * This can be run in a test environment to reset the state of the ports.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portStates hash of portName to state
+ */
+function setAllPortStates(node, vnid, portStates) {
```
Comment (dbooth-boston) [inline src/input-states.js:116]: good, i like it

File: src/input-states.js (hunk 6 above)
Comment (catch-point) [inline src/input-states.js:98]: I want to make it easy for the caller. If the port is not multi then socketId is undefined. Rather then requiring the caller to check for null socketId and have two different functions, I think it would be easier to do that here and just have one public function for both cases.

File: src/input-states.js (hunk 1 above)
Comment (jamesrdf) [inline src/input-states.js]: okay, but "this" is not used in this function, so it doesn't matter what it is.

File: src/input-states.js [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    return _.isFunction(port.isMulti);
+}
+
+/**
+ * Checks if a state already exists for this vnid and portName
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the port on this node
+ */
+function ensureVniStateExists(node, vnid, portName) {
```
Comment (dbooth-boston) [inline src/input-states.js:238]: Rename ensureVniStateExists can just be renamed to inputStateExists, since it is only checking to see if the state exists.

File: src/input-states.js (hunk 2 above)
Comment (catch-point) [inline src/input-states.js:18]: Only multi ports will have an array of states, non-multi ports will have just a state.

File: src/vni-manager.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+ *     var context = component.deleteVnis();
+ *     var context = component.deleteVni( vnid );
+ *     var context = nodeInstance.vni.delete();
+ *     var context = nodeInstance.vni().errorState( state );
+ *     var state = nodeInstance.vni().errorState();
+ *     var context = nodeInstance.vni().inputStates( {input: state} );
+ *     var inputStates = nodeInstance.vni().inputStates();
+ *     var context = nodeInstance.vni().outputState( state );
+ *     var state = nodeInstance.vni().outputState();
+ *     var context = nodeInstance.vni.delete();
```
Comment (catch-point) [inline src/vni-manager.js]: What's this?

File: src/vni-manager.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    _.extend( facade, 
+        {
+
+          /**
+           * Delete all vnis assiciated with the component node instance.
+           *
+           * @this component context
+           * @return component node instance context for easy chaining
+           */ 
+          deleteVnis: function() {
```
Comment (catch-point) [inline src/vni-manager.js]: What about deleteAllVnis? Sometimes the 's' can be overlooked and node.deleteVnis(vnid) might not do what one expects.

File: src/input-states.js [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+/**
+ * Changes the states of a port.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the port on this node
+ * @param stateArray a list of states for every attached socket on this port
+ */
+function setPortStateArray(node, vnid, portName, stateArray) {
```
Comment (dbooth-boston) [inline src/input-states.js:163]: Good.  This would normally only be used in test code though, because events (and states) are received one at a time -- not as an entire array at once.  It might be good to add  comment to that effect.

File: src/vni-manager.js [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+ * @this the component node instance unless otherwise specified on an API
+ * @param facade a component facade
+ * @return the component facade, extended to include the vni interface
+ * 
+ * Usage examples: 
+ *     var vni = this.nodeInstance.vni(); // retrieve an IIP vni 
+ *     var vni = this.nodeInstance.vni(vnid); // retrieve the vni associated with the vnid
+ *     var context = component.deleteVnis();
+ *     var context = component.deleteVni( vnid );
+ *     var context = nodeInstance.vni.delete();
```
Comment (catch-point) [inline src/vni-manager.js]: This should be var nodeInstance = nodeInstance.vni(vnid).delete(); to distinguish the vni facade context and the nodeInstance facade context.

File: src/input-states.js [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+function setAllPortStates(node, vnid, portStates) {
+    _.each(portStates, function(state, portName) {
+        setPortState(node, vnid, portName, state);
+    });
+    return node;
+}
+
+/**
+ * Changes the state of a port. When a new state is received on a port this
+ * should be called.
```
Comment (dbooth-boston) [inline src/input-states.js:125]: s/on a port/on a non-multi port/

And add: setPortStateBySocketId should be used to set the state on a multi/addressable port.

File: src/input-states.js [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+                if (_.isEmpty(node.vnis[vnid])) {
+                    delete node.vnis[vnid];
+                }
+            }
+        }
+    } else if (port.isMulti()) {
+        return setPortStateArray(node, vnid, portName, state);
+    } else {
+        if (!ensureVniStateExists(node, vnid, portName)) {
+            node.vnis = node.vnis || {};
```
Comment (dbooth-boston) [inline src/input-states.js:147]: This should call a common ensureVniExists(node, vnid) function instead of creating the object directly, so that all VNI creation is done in one place.

File: src/input-states.js [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        throw Error("This port is not addressable/multi: " + portName);
+    port.listAttached().forEach(function(socketId, i) {
+        setPortStateBySocketId(node, vnid, portName, socketId, stateArray[i]);
+    }, []);
+    return node;
+}
+
+/**
+ * Changes the state of a addressable/multi port socket. When a new state is received
+ * on a port this should be called.
```
Comment (dbooth-boston) [inline src/input-states.js:175]: s/on a port/on a multi port/

And add: setPortState should be called instead if the port is non-multi (non-addressable).

File: src/input-states.js [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the port on this node
+ * @param socketId the socketId pass to the data event handler
+ * @param state the new state of this port socket
+ */
+function setPortStateBySocketId(node, vnid, portName, socketId, state) {
+    if (socketId == null)
+        return setPortState(node, vnid, portName, state);
```
Comment (dbooth-boston) [inline src/input-states.js:184]: throw an error instead.  setPortState should have been called.

File: src/input-states.js (hunk 13 above)
Comment (catch-point) [inline src/input-states.js:125]: In a test environment, the port might be a multi port to reset all the states of the port.

File: src/input-states.js (hunk 16 above)
Comment (catch-point) [inline src/input-states.js:184]: I think socketId is important in this function, but not important at all to the caller. I think it that would make the caller's code harder to read to move the if statement was move there. It might also require this module to expose another public function (or variation), which we didn't want to do because it might make the module too confusing/busy.

File: src/input-states.js (hunk 1 above)
Comment (dbooth-boston) [inline src/input-states.js]: but the comment above seems to indicate that 'this' is returned

File: src/input-states.js (hunk 2 above)
Comment (dbooth-boston) [inline src/input-states.js:18]: oh, i see what you meant.  :)

File: src/vni-manager.js [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+ * File:  vni-manager.js
+ */
+var _ = require('underscore');
+
+var inputStates = require('./input-states');
+
+var IIP_VNID = '';
+
+/**
+ * Manage the node's vni's by vnid.  If no vnid is provided the IIP VNID will be used.
```
Comment (dbooth-boston) [inline src/vni-manager.js:11]: s/IIP VNID/default VNID '' (empty string)/

Explanation: the default VNID is used for regular nodes too -- not only for IIPs.

File: src/vni-manager.js [hunk 18]
```diff
@@ -0,0 +1,152 @@
+/**
+ * File:  vni-manager.js
+ */
+var _ = require('underscore');
+
+var inputStates = require('./input-states');
+
+var IIP_VNID = '';
```
Comment (dbooth-boston) [inline src/vni-manager.js]: s/IIP_VNID/DEFAULT_VNID/g

File: src/vni-manager.js [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+/**
+ * Manage the node's vni's by vnid.  If no vnid is provided the IIP VNID will be used.
+ * If the vni for the specified vnid does not exist, a new one will be created.
+ *
+ * @this the component node instance unless otherwise specified on an API
+ * @param facade a component facade
+ * @return the component facade, extended to include the vni interface
+ * 
+ * Usage examples: 
+ *     var vni = this.nodeInstance.vni(); // retrieve an IIP vni
```
Comment (dbooth-boston) [inline src/vni-manager.js]: s/retrieve an IIP vni/retrieve the default VNI/

File: src/vni-manager.js (hunk 9 above)
Comment (glenna-m) [inline src/vni-manager.js]: David had a requirement that we be able to delete a VNI.  I have provided two APIs to do it, though both share the same code underneath.  This one will delete the vni from the vnis array.

File: src/vni-manager.js (hunk 10 above)
Comment (glenna-m) [inline src/vni-manager.js]: Either is OK with me.  @dbooth-boston  - do you have a preference for the naming of this function:  deleteAllVnis or deleteVnis?

File: src/vni-manager.js (hunk 12 above)
Comment (glenna-m) [inline src/vni-manager.js]: OK.  Will make that change.

File: src/vni-manager.js [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+ * If the vni for the specified vnid does not exist, a new one will be created.
+ *
+ * @this the component node instance unless otherwise specified on an API
+ * @param facade a component facade
+ * @return the component facade, extended to include the vni interface
+ * 
+ * Usage examples: 
+ *     var vni = this.nodeInstance.vni(); // retrieve an IIP vni 
+ *     var vni = this.nodeInstance.vni(vnid); // retrieve the vni associated with the vnid
+ *     var context = component.deleteVnis();
```
Comment (dbooth-boston) [inline src/vni-manager.js]: This should be node.deleteAllVnis() -- not component.deleteAllVnis() -- because the VNIs are associated with the node (instances), not the component (factory).

Comment (dbooth-boston) [inline src/vni-manager.js]: s/context/node/ and s/component/node/  -- both are node (instances) AFAICT.  

In general I think it would be cleared to be more specific about what context is returned, such as "node" or "componentFacade" rather than just calling it "context".

File: src/vni-manager.js (hunk 12 above)
Comment (dbooth-boston) [inline src/vni-manager.js]: And s/nodeInstance/node/

File: src/vni-manager.js (hunk 20 above)
Comment (glenna-m) [inline src/vni-manager.js]: It is the node instance we are working with.  Will rename.

File: src/vni-manager.js [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+
+var inputStates = require('./input-states');
+
+var IIP_VNID = '';
+
+/**
+ * Manage the node's vni's by vnid.  If no vnid is provided the IIP VNID will be used.
+ * If the vni for the specified vnid does not exist, a new one will be created.
+ *
+ * @this the component node instance unless otherwise specified on an API
```
Comment (dbooth-boston) [inline src/vni-manager.js]: General comment: There are several places in the code where I found the names confusing.  I think the naming throughout this file should be aligned with the naming conventions that we decided last week:
https://docs.google.com/document/d/1mb_8DCV7nnvhYwc0Ns_gCWpzyXcM1_8qs2u9GJpIZP4/edit#
- An instance of a NoFlo component will be called a "node" (not nodeInstance);
- A NoFlo component (actually a node factory) will be called a "component".  
  - I don't know if we decided what to call the component facade, but how about "componentFacade"?

File: src/vni-manager.js [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+ */
+module.exports = function( facade ) { 
+
+    _.extend( facade, 
+        {
+
+          /**
+           * Delete all vnis assiciated with the component node instance.
+           *
+           * @this component context
```
Comment (dbooth-boston) [inline src/vni-manager.js]: s/component/node/

File: src/vni-manager.js (hunk 21 above)
Comment (glenna-m) [inline src/vni-manager.js]: Did not see this before my last commit.  So if I understand write, you want the following mappings:
nodeInstance -> node
facade -> componentFacade.  

If so, will do that now.

Comment (dbooth-boston) [inline src/vni-manager.js]: yes, and avoid using the word "component" when talking about the node instance.  Use "node" instead.  thanks

File: src/vni-manager.js [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * Manage the node's vni's by vnid.  If no vnid is provided the IIP VNID will be used.
+ * If the vni for the specified vnid does not exist, a new one will be created.
+ *
+ * @this the component node instance unless otherwise specified on an API
+ * @param facade a component facade
+ * @return the component facade, extended to include the vni interface
+ * 
+ * Usage examples: 
+ *     var vni = this.nodeInstance.vni(); // retrieve the default VNI
```
Comment (dbooth-boston) [inline src/vni-manager.js]: AFAICT this should be:

<pre>
var vni = node.vni(); // retrieve the default VNI 
</pre>

File: src/vni-manager.js [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+ * Manage the node's vni's by vnid.  If no vnid is provided the IIP VNID will be used.
+ * If the vni for the specified vnid does not exist, a new one will be created.
+ *
+ * @this the component node instance unless otherwise specified on an API
+ * @param facade a component facade
+ * @return the component facade, extended to include the vni interface
+ * 
+ * Usage examples: 
+ *     var vni = this.nodeInstance.vni(); // retrieve the default VNI 
+ *     var vni = this.nodeInstance.vni(vnid); // retrieve the vni associated with the vnid
```
Comment (dbooth-boston) [inline src/vni-manager.js]: Similarly this should be:

<pre>
var vni = node.vni(vnid); // retrieve the vni associated with the vnid
</pre>

File: src/vni-manager.js [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+ *
+ *     component.deleteAllVnis();
```
Comment (dbooth-boston) [inline src/vni-manager.js]: s/component/node/ and in several more places.

File: src/vni-manager.js [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+ *     component.deleteVni( vnid );
+ *     nodeInstance.vni(vnid).delete();
```
Comment (dbooth-boston) [inline src/vni-manager.js]: s/nodeInstance/node/  and similarly in several more places.

File: src/vni-manager.js [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+    _.extend( facade, 
+        {
+
+          /**
+           * Delete all vnis assiciated with the node instance.
+           *
+           * @this node instance
+           * @return node instance nodeInstance for easy chaining
+           */ 
+          deleteAllVnis: function() {
```
Comment (dbooth-boston) [inline src/vni-manager.js:47]: good

File: src/vni-manager.js [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+
+              return {
+                  delete: _.bind( this.deleteVni, this, vnid ),
+                  inputStates: _.partial( inputStates, this, vnid ), 
+                  errorState: _.partial( errorState, this.vnis[vnid] ), 
+                  outputState: _.partial( outputState, this.vnis[vnid] ) 
+              };
+          },
+
+          vnis: {}
```
Comment (dbooth-boston) [inline src/vni-manager.js:100]: I am confused about whether this module returns a _component_ facade or a _node_ (instance).  The vnis hash needs to be stored on the _node_ -- not the component facade.  Maybe you can explain on a hangout.

File: src/vni-manager.js [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+    }); // extend
+} // module export
+
+/**
+ * Get/Set error state on the vni 
+ * 
+ * @this vni context 
+ * @param state error state to be set
+ */
+function errorState( vni, state ) {
```
Comment (dbooth-boston) [inline src/vni-manager.js:115]: good

File: src/vni-manager.js [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+/**
+ * Get/Set output state on the vni 
+ * 
+ * @this vni context 
+ * @param state output state to be set
+ */
+function outputState( vni, state ) {
+
```
Comment (dbooth-boston) [inline src/vni-manager.js:146]: good

File: test/vni-manager-mocha.js [hunk 31]
```diff
@@ -0,0 +1,306 @@
...
+
+var test = require('./common-test');
+
+var vniManager = require('../src/vni-manager');
+
+describe('vni-manager', function() {
+
+    beforeEach(function() {
+
+        component = test.createComponent(componentFactory({
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js]: s/component/node/

File: test/vni-manager-mocha.js [hunk 32]
```diff
@@ -0,0 +1,306 @@
...
+        component.deleteAllVnis();
+    });
+
+    it('should exist as a function', function() {
+      vniManager.should.exist;
+      vniManager.should.be.a('function');
+    });
+
+    describe('#vnis', function() {
+        it('should exist as an object in the component instance', function() {
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js]: s/component instance/node (instance)/ and throughout this file.

File: test/vni-manager-mocha.js [hunk 33]
```diff
@@ -0,0 +1,306 @@
...
+            component.vnis.should.exist;
+            component.vnis.should.be.a('object');
+            Object.keys( component.vnis ).should.have.length( 0 );
+        });
+    });
+
+    describe('#vni', function() {
+
+        it('should create and return a new IIP vni if no vnid is specified and no vni exists', function() {
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js]: s/IIP/default/

File: test/vni-manager-mocha.js [hunk 34]
```diff
@@ -0,0 +1,306 @@
...
+            testVni.should.be.an('object');
+            testVni.inputStates.should.exist;
+            testVni.inputStates.should.be.a('function');
+            testVni.errorState.should.exist;
+            testVni.errorState.should.be.a('function');
+            testVni.outputState.should.exist;
+            testVni.outputState.should.be.a('function');
+
+            expect( testVni.parentVni ).to.be.undefined;
+            expect( testVni.previousLms ).to.be.undefined;
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js]: The previousLms would be a property of the output or error states -- not directly in the VNI.

File: test/vni-manager-mocha.js [hunk 35]
```diff
@@ -0,0 +1,306 @@
...
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js]: ditto

File: test/vni-manager-mocha.js [hunk 36]
```diff
@@ -0,0 +1,306 @@
...
+            it('should delete the VNI', function() {
+
+                Object.keys( component.vnis ).should.have.length( 0 );
+
+                var testVnid = '';
+                var testVni = component.vni( testVnid );
+                testVni.should.be.an('object');
+                Object.keys( component.vnis ).should.have.length( 1 );
+
+                var context = testVni.delete();
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js]: s/context/node/   and several other places below

File: test/vni-manager-mocha.js [hunk 37]
```diff
@@ -0,0 +1,306 @@
...
+                var context = testVni.outputState( state );
+                context.should.be.an('object');
+
+                // Test clearing the state 
+                var clearContext = testVni.outputState( undefined );
+                clearContext.should.be.an('object');
+
+                // Test get of cleared state 
+                var clearedState = testVni.outputState();
+                expect( clearedState ).to.be.an('undefined');
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js:319]: good

File: src/vni-manager.js (hunk 28 above)
Comment (glenna-m) [inline src/vni-manager.js:100]: Added a test that demonstrates that the function results in an extended component facade with the new vni-manager apis.

File: test/vni-manager-mocha.js [hunk 38]
```diff
@@ -0,0 +1,377 @@
...
+                    ondata: promiseOutput(function(payload){
+                        return payload + " world2";
+                    })
+                }},
+                outPorts: promiseOutput.outPorts
+              }, vniManager )
+            );
+ 
+            node1.should.not.equal(node2);
+            node1.vnis.should.not.equal(node2.vnis);
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js:84]: good, thanks

**Comment by dbooth-boston** [comment]
LGTM.  thanks!

File: src/vni-manager.js [hunk 39]
```diff
@@ -0,0 +1,158 @@
...
+        if ( _.isUndefined(state) ) { 
+            // Clear state
+            delete vni.errorState;
+        } else { 
+            // Set state
+            vni.errorState = state;
+        }
+
+        return node;
```
Comment (catch-point) [inline src/vni-manager.js]: I think this should be 'return this;' so that it returns the vni facade and not the node facade. Same goes for outputState below.

File: src/vni-manager.js [hunk 40]
```diff
@@ -0,0 +1,158 @@
...
+
+              vnid = vnid || DEFAULT_VNID;
+
+              if ( _.isUndefined( this.vnis[vnid] )  ) {
+
+                  // have no vni so create one
+                  this.vnis[vnid] = {}; 
+
+                  // TODO: Add parentVni setting here
+                  // TODO: Add previousLms here
```
Comment (dbooth-boston) [inline src/vni-manager.js]: Not previousLms.  They are a part of the state objects -- not the VNI directly.

File: test/vni-manager-mocha.js [hunk 41]
```diff
@@ -0,0 +1,377 @@
...
+
+var test = require('./common-test');
+
+var vniManager = require('../src/vni-manager');
+
+describe("vni-manager", function() {
+
+    beforeEach(function() {
+
+        node = test.createComponent(componentFactory({
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js:24]: Should we rename createComponent to createNode?  I'm getting confused about which objects are components (i.e., node factories) and which are nodes (i.e., node instances).

**Comment by catch-point** [comment]
LGTM

File: test/vni-manager-mocha.js [hunk 42]
```diff
@@ -0,0 +1,377 @@
...
+        // clean up
+        node.deleteAllVnis();
+    });
+
+    it("should exist as a function", function() {
+        vniManager.should.exist;
+        vniManager.should.be.a('function');
+    });
+
+    it("should return a component facade", function() {
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js:46]: Should this say "node facade" instead of "component facade"?  I'm still a bit confused about which objects are nodes and which are components, so if we can be consistent in our naming it will help.

File: test/vni-manager-mocha.js [hunk 43]
```diff
@@ -0,0 +1,377 @@
+/**
+ * File: vni-manager-mocha.js
+ * Unit tests for the vni APIs defined in src/vni-manager.js
+ */
+
+var chai = require('chai');
+var expect = chai.expect;
+var should = chai.should();
+
+var componentFactory = require('../src/noflo-component-factory');
```
Comment (dbooth-boston) [inline test/vni-manager-mocha.js:10]: Let's rename noflo-component-factory --> noflo-node-factory
and componentFactory --> nodeFactory

But we can do it in a separate PR

**Comment by dbooth-boston** [comment]
LGTM!

File: test/vni-manager-mocha.js (hunk 42 above)
Comment (catch-point) [inline test/vni-manager-mocha.js:46]: My take away from yesterdays discussion on this is that the term node should be used when referencing a particular instance and the term component should be used when referencing a class/collection of them.
